File: find/max_subarray.py
from random import randint
from math import inf


def max_crossing(arr, l, m, h):
    lind = m
    lsum = -inf
    sum_ = 0

    for i in range(m, l - 1, -1):
        sum_ += arr[i]
        if sum_ > lsum:
            lsum = sum_
            lind = i

    rind = m
    rsum = -inf
    sum_ = 0

    for i in range(m + 1, h):
        sum_ += arr[i]
        if sum_ > rsum:
            rsum = sum_
            rind = i

    return lind, rind, rsum + lsum


def max_subarray(arr, l, h):
    if l != h:
        mid = (l + h) // 2

        ll, lh, lsum = max_subarray(arr, l, mid)
        rl, rh, rsum = max_subarray(arr, mid + 1, h)
        cl, ch, csum = max_crossing(arr, l, mid, h + 1)
        
        if lsum >= rsum and lsum >= csum:
            return ll, lh, lsum
        elif rsum >= lsum and rsum >= csum:
            return rl, rh, rsum
        else:
            return cl, ch, csum
    else:
        return l, h, arr[l]


def main():
    arr = [randint(-10, 10) for x in range(10)]
    print("ARRAY: ", arr, "\nRESULT:", max_subarray(arr, 0, len(arr) - 1))


if __name__ == "__main__":
    main()

File: search/max_subarray.hs
maxCrossing :: Ord a => [a] -> (Int, Int, a)
maxCrossing xs = (l, h, lsum + rsum)
    where m = length xs `div` 2
          (l, lsum) = foldr1 (\acc x -> acc) (l, 999) $ take m xs
          (h, rsum) = foldl1 (\acc x -> acc) (h, 999) $ drop m xs


maxSubarray :: Ord a => [a] -> (Int, Int, a)
maxSubarray xs
    | lsum >= rsum && lsum >= csum = (ll, lh, lsum)
    | rsum >= lsum && rsum >= csum = (rl, rh, rsum)
    | otherwise = (cl, ch, csum)
    where 
        mid = length xs `div` 2
        (ll, lh, lsum) = maxSubarray $ take mid xs
        (rl, rh, rsum) = maxSubarray $ drop mid xs
        (cl, ch, csum) = maxCrossing xs


main :: IO ()
main = do
    print $ maxSubarray [-9, 7, 2, 6, 7, 9, -5, -10, 6, -6] 
